File: sumRootToLeaf.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @return {number}
 */
var sumNumbers = function(root) {
    if(root == null) return 0;
    let result = 0;
    let currSum = 0;
    let stack = [];
    while(root != null || stack.length ){
        while(root!=null){
            currSum = currSum*10 + root.val;
            stack.push([root,currSum]);
            root = root.left;
        }
        let tuple = stack.pop();
         root = tuple[0];
        currSum = tuple[1];
        if(root.left == null && root.right == null) {
            result = result + currSum;
        }
        root = root.right;
        
    }
    return result;
};

//time complexity : O(n)
//space complexity : O(h)-height of the tree

File: sumroot_to_leaf.py
# Time Complexity : O(N) 
# Space Complexity :    O(H)  - height of tree
# Did this code successfully run on Leetcode : YES
# Any problem you faced while coding this : NO


# Definition for a binary tree node.
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right
        
class Solution:
    def sumNumbers(self, root: TreeNode) -> int:
        if root==None:
            return 0
        self.ans = 0
        self.helper(root,0)
        return self.ans
        
        
    def helper(self,root,currSum):
        if root==None:
            return
        
        #find currSum at the node
        currSum = currSum*10+root.val
        
        self.helper(root.left,currSum)
        
        #if no left and right child add cursum to the global result
        if root.left==None and root.right==None:
            self.ans+=currSum
        
        self.helper(root.right,currSum)
        
            
            
            
                
        